src/lib.rs

```diff
@@ -103,13 +103,74 @@ pub enum Instruction {
     ADD(usize),
 }
 
-#[derive(Debug)]
+impl rlp::Encodable for Instruction {
+    fn rlp_append(&self, s: &mut rlp::RlpStream) {
+        match self {
+            Instruction::EXTENSION(ref ext) => s.begin_list(2).append(&3usize).append_list(ext),
+            Instruction::BRANCH(size) => s.begin_list(2).append(&0usize).append(size),
+            Instruction::HASHER(size) => s.begin_list(2).append(&1usize).append(size),
+            Instruction::LEAF(size) => s.begin_list(2).append(&2usize).append(size),
+            Instruction::ADD(index) => s.begin_list(2).append(&4usize).append(index),
+        };
+    }
+}
+
+impl rlp::Decodable for Instruction {
+    fn decode(rlp: &rlp::Rlp) -> Result<Self, rlp::DecoderError> {
+        let instrrlp = rlp.at(0usize)?;
+        let instr: usize = instrrlp.as_val()?;
+
+        if instr >= 5 {
+            return Err(rlp::DecoderError::Custom("Invalid instruction opcode {}"));
+        }
+
+        if instr != 3 {
+            let size: usize = rlp.at(1usize)?.as_val()?;
+            let i = match instr {
+                0 => Instruction::BRANCH(size),
+                1 => Instruction::HASHER(size),
+                2 => Instruction::LEAF(size),
+                4 => Instruction::ADD(size),
+                _ => panic!("This should never happen!"), /* Famous last words */
+            };
+
+            return Ok(i);
+        }
+
+        let ext = rlp.at(1usize)?.as_list()?;
+        Ok(Instruction::EXTENSION(ext))
+    }
+}
+
+#[derive(Debug, PartialEq)]
 pub struct Multiproof {
     pub hashes: Vec<Vec<u8>>,           // List of hashes in the proof
     pub instructions: Vec<Instruction>, // List of instructions in the proof
     pub keyvals: Vec<Vec<u8>>,          // List of RLP-encoded (key, value) pairs in the proof
 }
 
+impl rlp::Encodable for Multiproof {
+    fn rlp_append(&self, s: &mut rlp::RlpStream) {
+        s.begin_list(2);
```

Comment:
I'm not sure about the RLP library, but shouldn't this be `s.begin_list(3);`?

Reply:
it should, which is weird because it works and it shouldn't :thinking: